istioctl/cmd/istioctl/main.go

```diff
@@ -812,3 +813,23 @@ func handleNamespaces(objectNamespace string) (string, error) {
 	}
 	return defaultNamespace, nil
 }
+
+// BuildClientConfig is a helper function that builds client config from a master
+// url or a kubeconfig filepath.
+// This is a modified version of k8s.io/client-go/tools/clientcmd/BuildConfigFromFlags with the
+// difference that the kubeconfigPath can also contain multiple paths representing precedences
+// of kubernetes config paths.
+func BuildClientConfig(masterURL, kubeconfigPath string) (*rest.Config, error) {
```

Comment:
I can but the difficult part is that it can only be correctly tested with at least two valid kuberenetes configs. While easy to set locally I'm afraid it won't be as easy running the test on CI. Got an idea?

Reply:
Can you add the ability to read from a particular context as well ? It’s common for kube config to have multiple contexts and last I checked there was no way to pick a context. It was using the default context.

Reply:
I found that there is no test file for `main.go` now. Yes, I think we can go ahead with this PR and create new test file for istioctl command in future.

Reply:
Thanks @gyliu513 
@rshriram I will create a new issue for the support of specifying a context. Probably adding a `--kubecontext` string flag

Reply:
https://github.com/istio/istio/issues/4081 and the PR title indicate the error is with `KUBECONFIG` supporting multiple configs, but this code is addressing multiple files specified by `--kubeconfig` flag. Something like the following might work on line 831 given `NewDefaultClientConfigLoadingRules` handles KUBECONFIG case.

```golang
loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()                                                                                                                                                                                                                                                                                                           
configOverrides := &clientcmd.ConfigOverrides{}                                                                                                                                                                                                                                                                                          
kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides)                                                                                                                                                                                                                                                                      
config, err := kubeConfig.ClientConfig()
```

Reply:
@ayj Thanks. It's not accurate as the `getRealKubeConfig` function will get the env variable `KUBECONFIG` if the user didn't specify the `--kubeconfig`. The snippet you mentioned is ignoring that flag if exists and just use the system's.

I tried to minimize the amount of changes not to break anything as we don't currently test `istioctl`. Anyway, I will refactor it a little to try and make it more clear.